File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/inference/OpenRouterOAuth.kt
package org.wyrdsekai.app.inference

import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.random.Random
import org.wyrdsekai.app.platform.percentEncode
import org.wyrdsekai.app.platform.sha256
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

/**
 * OpenRouter OAuth PKCE flow for automated API key acquisition.
 *
 * Flow:
 * 1. Generate code_verifier + code_challenge (S256)
 * 2. Open browser to openrouter.ai/auth?callback_url=...&code_challenge=...
 * 3. User logs in, redirected back with ?code=AUTH_CODE
 * 4. Exchange code for API key via POST /api/v1/auth/keys
 *
 * Callback constraint: HTTPS on port 443 or 3000, or http://localhost:3000
 */
object OpenRouterOAuth {

    private const val AUTH_URL = "https://openrouter.ai/auth"
    private const val EXCHANGE_URL = "https://openrouter.ai/api/v1/auth/keys"

    /**
     * Default callback URL for phone clients. OpenRouter only permits
     * https:443, https:3000, or http://localhost:3000 — no custom URI
     * schemes. We use the loopback variant and intercept the redirect
     * inside our in-app WebView before it actually tries to reach
     * localhost. See OpenRouterAuthScreen.
     */
    const val LOOPBACK_CALLBACK = "http://localhost:3000/callback"

    private val json = Json { ignoreUnknownKeys = true }

    /** PKCE state — held between redirect and callback. */
    data class PkceState(
        val codeVerifier: String,
        val codeChallenge: String,
    )

    private var pendingState: PkceState? = null

    /**
     * Build the authorization URL to open in the browser.
     * @param callbackUrl Where OpenRouter redirects after auth (e.g. "https://yourapp.com/callback")
     * @return The full URL to open, and the PKCE state to hold for exchange.
     */
    fun buildAuthUrl(callbackUrl: String): Pair<String, PkceState> {
        val verifier = generateCodeVerifier()
        val challenge = computeS256Challenge(verifier)
        val state = PkceState(verifier, challenge)
        pendingState = state

        val url = "$AUTH_URL?callback_url=${encodeUrl(callbackUrl)}" +
            "&code_challenge=$challenge" +
            "&code_challenge_method=S256"

        return url to state
    }

    /**
     * Exchange the authorization code for an API key.
     * @param code The code from the callback URL query parameter.
     * @param pkceState The PKCE state from buildAuthUrl.
     * @return The API key string, or null on failure.
     */
    suspend fun exchangeCode(code: String, pkceState: PkceState? = pendingState): OAuthResult {
        val state = pkceState ?: return OAuthResult(null, "No PKCE state — did you call buildAuthUrl first?")

        val client = HttpClient()
        return try {
            val response = client.post(EXCHANGE_URL) {
                contentType(ContentType.Application.Json)
                setBody("""{"code":"$code","code_verifier":"${state.codeVerifier}","code_challenge_method":"S256"}""")
            }

            if (response.status.value == 200) {
                val body = json.decodeFromString<ExchangeResponse>(response.bodyAsText())
                pendingState = null
                OAuthResult(body.key, null)
            } else {
                val body = response.bodyAsText()
                OAuthResult(null, "Exchange failed (${response.status.value}): $body")
            }
        } catch (e: Exception) {
            OAuthResult(null, "Exchange error: ${e.message}")
        } finally {
            client.close()
        }
    }

    data class OAuthResult(val key: String?, val error: String?)

    @Serializable
    private data class ExchangeResponse(val key: String? = null)

    // ── PKCE helpers ─────────────────────────────────────────────────

    private fun generateCodeVerifier(): String {
        val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
        return (1..64).map { chars[Random.nextInt(chars.length)] }.joinToString("")
    }

    /**
     * S256 challenge: base64url(sha256(verifier)).
     * Uses platform expect/actual for SHA-256 — falls back to plain if unavailable.
     */
    private fun computeS256Challenge(verifier: String): String {
        return try {
            base64UrlEncode(sha256(verifier.encodeToByteArray()))
        } catch (_: Exception) {
            // Fallback: plain (less secure but functional)
            verifier
        }
    }

    @OptIn(ExperimentalEncodingApi::class)
    private fun base64UrlEncode(bytes: ByteArray): String =
        Base64.UrlSafe.withPadding(Base64.PaddingOption.ABSENT).encode(bytes)

    private fun encodeUrl(url: String): String = percentEncode(url)
}

File: clients/rn/src/inference/OpenRouterOAuth.ts
/**
 * OpenRouter OAuth PKCE flow — RN port.
 *
 * Mirrors the API of {@code OpenRouterOAuth.kt} (KMP) and
 * {@code OpenRouterOAuth.java} (server). See those for the full design.
 *
 * Flow:
 *   1. buildAuthUrl(LOOPBACK_CALLBACK) → {authUrl, pkce}
 *   2. Render authUrl inside a WebView (OpenRouterAuthScreen).
 *   3. WebView intercepts the {@link LOOPBACK_CALLBACK} navigation —
 *      the localhost URL never actually hits the network.
 *   4. Pull {@code ?code=...} from the intercepted URL, call exchangeCode.
 *   5. exchangeCode POSTs to OpenRouter and returns the API key.
 *
 * OpenRouter only permits https:443, https:3000, or http://localhost:3000
 * as callback URLs — explicitly no custom URI schemes. The loopback URL
 * doesn't need a listener; the WebView's URL-intercept aborts the load
 * before any TCP connection is attempted.
 */
import { sha256 } from 'js-sha256';

const AUTH_URL = 'https://openrouter.ai/auth';
const EXCHANGE_URL = 'https://openrouter.ai/api/v1/auth/keys';

/** Default callback for phone clients (intercepted in WebView). */
export const LOOPBACK_CALLBACK = 'http://localhost:3000/callback';

/** PKCE state — held between redirect and exchange. */
export interface PkceState {
  codeVerifier: string;
  codeChallenge: string;
}

export interface OAuthResult {
  key: string | null;
  error: string | null;
}

const VERIFIER_CHARS =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';

function generateCodeVerifier(): string {
  const buf = new Uint8Array(64);
  if (
    typeof globalThis !== 'undefined' &&
    (globalThis as any).crypto?.getRandomValues
  ) {
    (globalThis as any).crypto.getRandomValues(buf);
  } else {
    // Should not happen — index.js imports react-native-get-random-values.
    for (let i = 0; i < buf.length; i++) {
      buf[i] = Math.floor(Math.random() * 256);
    }
  }
  let out = '';
  for (let i = 0; i < buf.length; i++) {
    out += VERIFIER_CHARS[buf[i] % VERIFIER_CHARS.length];
  }
  return out;
}

function base64UrlEncodeBytes(bytes: ArrayLike<number>): string {
  let bin = '';
  for (let i = 0; i < (bytes as any).length; i++) {
    bin += String.fromCharCode((bytes as any)[i]);
  }
  // RN doesn't ship Buffer; `btoa` is provided by react-native's globals.
  const b64 =
    typeof btoa !== 'undefined'
      ? btoa(bin)
      : // Fallback — should not be reached in RN
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        (require('buffer').Buffer as typeof import('buffer').Buffer)
          .from(bin, 'binary')
          .toString('base64');
  return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

function computeS256Challenge(verifier: string): string {
  // js-sha256 returns an ArrayBuffer when called with `.arrayBuffer()`.
  const digest = sha256.arrayBuffer(verifier);
  return base64UrlEncodeBytes(new Uint8Array(digest));
}

/**
 * Build the authorization URL to load in the WebView.
 * @returns [authUrl, pkceState]. Keep pkceState alive until exchangeCode.
 */
export function buildAuthUrl(callbackUrl: string): {
  authUrl: string;
  pkce: PkceState;
} {
  const verifier = generateCodeVerifier();
  const challenge = computeS256Challenge(verifier);
  const url =
    `${AUTH_URL}?callback_url=${encodeURIComponent(callbackUrl)}` +
    `&code_challenge=${challenge}` +
    `&code_challenge_method=S256`;
  return { authUrl: url, pkce: { codeVerifier: verifier, codeChallenge: challenge } };
}

/**
 * Exchange the authorization code for an API key.
 * Posts to {@code https://openrouter.ai/api/v1/auth/keys} with the verifier.
 */
export async function exchangeCode(
  code: string,
  pkce: PkceState,
): Promise<OAuthResult> {
  try {
    const res = await fetch(EXCHANGE_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        code,
        code_verifier: pkce.codeVerifier,
        code_challenge_method: 'S256',
      }),
    });
    if (!res.ok) {
      const body = await res.text().catch(() => '');
      return { key: null, error: `Exchange failed (${res.status}): ${body}` };
    }
    const body = (await res.json()) as { key?: string };
    if (!body.key) {
      return { key: null, error: 'OpenRouter response missing key' };
    }
    return { key: body.key, error: null };
  } catch (e: any) {
    return { key: null, error: `Exchange error: ${e?.message ?? String(e)}` };
  }
}

/** Pull the {@code code} query param out of the intercepted callback URL. */
export function parseCodeFromCallbackUrl(url: string): string | null {
  const q = url.indexOf('?') >= 0 ? url.slice(url.indexOf('?') + 1) : '';
  if (!q) return null;
  for (const kv of q.split('&')) {
    const eq = kv.indexOf('=');
    if (eq > 0 && kv.slice(0, eq) === 'code') {
      return decodeURIComponent(kv.slice(eq + 1));
    }
  }
  return null;
}
